x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/histogram/HistogramPercentile.java

```diff
@@ -0,0 +1,131 @@
+/**
+ * Extracts a percentile value from a single histogram value.
+ * Note that this function is currently only intended for usage in surrogates and not available as a user-facing function.
+ * Therefore, it is intentionally not registered in {@link org.elasticsearch.xpack.esql.expression.function.EsqlFunctionRegistry}.
+ */
+public class HistogramPercentile extends EsqlScalarFunction {
+
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(
+        Expression.class,
+        "HistogramPercentile",
+        HistogramPercentile::new
+    );
+
+    private final Expression histogram;
+    private final Expression percentile;
+
+    @FunctionInfo(returnType = { "double" })
+    public HistogramPercentile(
+        Source source,
+        @Param(name = "histogram", type = { "exponential_histogram" }) Expression histogram,
+        @Param(name = "percentile", type = { "double", "integer", "long", "unsigned_long" }) Expression percentile
+    ) {
+        super(source, List.of(histogram, percentile));
+        this.histogram = histogram;
+        this.percentile = percentile;
+    }
+
+    private HistogramPercentile(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(Expression.class));
+    }
+
+    Expression histogram() {
+        return histogram;
+    }
+
+    Expression percentile() {
+        return percentile;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        return isType(histogram, dt -> dt == DataType.EXPONENTIAL_HISTOGRAM, sourceText(), DEFAULT, "exponential_histogram").and(
+            isType(percentile, DataType::isNumeric, sourceText(), DEFAULT, "numeric types")
+        );
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataType.DOUBLE;
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new HistogramPercentile(source(), newChildren.get(0), newChildren.get(1));
+    }
+
+    @Override
+    public boolean foldable() {
+        return histogram.foldable() && percentile.foldable();
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, HistogramPercentile::new, histogram, percentile);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(histogram);
+        out.writeNamedWriteable(percentile);
+    }
+
+    @Evaluator(warnExceptions = ArithmeticException.class)
+    static void process(DoubleBlock.Builder resultBuilder, ExponentialHistogram value, double percentile) {
+        if (percentile < 0.0 || percentile > 100.0) {
+            throw new ArithmeticException("Percentile value must be in the range [0, 100], got: " + percentile);
```

Comment:
Question: Is it a good or a bad practice to include the invalid value in the warning message? E.g. `Asin` doesn't include the wrong value in its warnings.

Reply:
I think we should reject this during verification instead of checking every row.

Reply:
If the percentile is a constant with the integration, we should verify it and reject invalid values instead. But this should be okay now.